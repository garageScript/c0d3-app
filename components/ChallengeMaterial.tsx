import React, { useState } from 'react'
import {
  Challenge,
  ChallengeSubmissionData,
  UserSubmission,
  UserSubmissionsObject
} from '../@types/challenge'
import { LessonStatus } from '../@types/lesson'
import NavLink from './NavLink'
import Markdown from 'markdown-to-jsx'
import gitDiffParser, { File } from 'gitdiff-parser'
import ReactDiffViewer from 'react-diff-viewer'
import Prism from 'prismjs'
import dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime'
import { GiveStarCard } from '../components/GiveStarCard'
import _ from 'lodash'

dayjs.extend(relativeTime)

type CurrentChallengeID = string | null

type ReviewStatusProps = {
  status: string
  reviewerUserName: string | null
}

export const ReviewStatus: React.FC<ReviewStatusProps> = ({
  status,
  reviewerUserName
}) => {
  let reviewStatusComment
  let statusClassName
  const profileLink = (
    <NavLink
      className="text-reset"
      path="/profile/[username]"
      as={`/profile/${reviewerUserName}`}
    >
      {reviewerUserName}
    </NavLink>
  )
  switch (status) {
    case 'passed':
      reviewStatusComment = (
        <>Your solution was reviewed and accepted by {profileLink}</>
      )
      statusClassName = 'border border-success text-success'
      break
    case 'needMoreWork':
      reviewStatusComment = (
        <>Your solution was reviewed and rejected by {profileLink}</>
      )
      statusClassName = 'border border-danger text-danger'
      break
    case 'open':
      reviewStatusComment = (
        <>Your submission is currently waiting to be reviewed</>
      )
      statusClassName = 'border border-warning text-warning'
      break
    default:
      return null
  }
  return (
    <div className={`text-center p-2 my-2 ${statusClassName}`}>
      {reviewStatusComment}
    </div>
  )
}

type StatusIconProps = {
  status: string
}
const StatusIcon: React.FC<StatusIconProps> = ({ status }) => {
  if (status === 'unsubmitted') {
    return null
  }
  let statusIconUrl
  switch (status) {
    case 'passed':
      statusIconUrl = '/assets/curriculum/icons/checkmark.svg'
      break
    case 'needMoreWork':
      statusIconUrl = '/assets/curriculum/icons/rejected.svg'
      break
    case 'open':
      statusIconUrl = '/assets/curriculum/icons/pending.svg'
  }
  return <img width="25px" height="25px" src={statusIconUrl} />
}

type ChallengeTitleCardProps = {
  key: string
  id: string
  title: string
  challengeNum: number
  submissionStatus: string
  active?: boolean
  setCurrentChallenge: React.Dispatch<CurrentChallengeID>
}

export const ChallengeTitleCard: React.FC<ChallengeTitleCardProps> = ({
  submissionStatus,
  active,
  title,
  id,
  challengeNum,
  setCurrentChallenge
}) => {
  const cardStyles: string[] = ['challenge-title-card']
  if (active) {
    cardStyles.push('challenge-title-card--active')
  }
  if (submissionStatus === 'passed') {
    cardStyles.push('challenge-title-card--done')
  } else {
    cardStyles.push('shadow-sm', 'border-0')
  }

  return (
    <div
      data-testid="challenge-title"
      className={`card mb-2 ${cardStyles.join(' ')}`}
      onClick={() => setCurrentChallenge(id)}
    >
      <div className="card-body d-flex justify-content-between">
        <div>{`${challengeNum}. ${title}`}</div>
        <StatusIcon status={submissionStatus} />
      </div>
    </div>
  )
}

type ChallengeQuestionCardProps = {
  currentChallenge: ChallengeSubmissionData
}
export const ChallengeQuestionCard: React.FC<ChallengeQuestionCardProps> = ({
  currentChallenge
}) => {
  const diff = _.get(currentChallenge, 'submission.diff', '')
  const comment = _.get(currentChallenge, 'submission.comment', '')
  const updatedAt = _.get(currentChallenge, 'submission.updatedAt', Date.now())
  const reviewerUserName = _.get(
    currentChallenge,
    'submission.reviewer.username',
    null
  )
  let files = null

  if (diff) files = gitDiffParser.parse(diff)

  const renderFile = ({ hunks, newPath }: File) => {
    const oldValue: String[] = []
    const newValue: String[] = []

    hunks.forEach(hunk => {
      hunk.changes.forEach(change => {
        if (change.isDelete) oldValue.push(change.content)
        else if (change.isInsert) newValue.push(change.content)
        else {
          oldValue.push(change.content)
          newValue.push(change.content)
        }
      })
    })

    const syntaxHighlight = (str: string): any => {
      if (!str) return

      const language = Prism.highlight(
        str,
        Prism.languages.javascript,
        'javascript'
      )
      return <span dangerouslySetInnerHTML={{ __html: language }} />
    }

    return (
      <ReactDiffViewer
        key={_.uniqueId()}
        oldValue={oldValue.join('\n')}
        newValue={newValue.join('\n')}
        renderContent={syntaxHighlight}
        splitView={false}
        leftTitle={`${newPath}`}
      />
    )
  }

  return (
    <>
      <div className="card shadow-sm border-0">
        <div className="card-body challenge-question">
          <h1 data-testid="challenge-question-title" className="card-title">
            {currentChallenge.title}
          </h1>
          <div
            data-testid="challenge-question-description"
            className="bg-light p-3 mt-3"
          >
            <Markdown>{currentChallenge.description}</Markdown>
          </div>
        </div>
      </div>

      {diff && (
        <div className="card shadow-sm border-0 mt-3">
          <div className="card-header bg-white">
            Submitted {dayjs(parseInt(updatedAt)).fromNow()}
          </div>
          <div className="card-body">
            <div className="rounded-lg overflow-hidden">
              {files && files.map(renderFile)}
            </div>
          </div>
          <div className="card-footer bg-white">
            {comment && <Markdown>{comment}</Markdown>}
            <ReviewStatus
              status={currentChallenge.status}
              reviewerUserName={reviewerUserName}
            />
          </div>
        </div>
      )}
    </>
  )
}

type ChallengesCompletedCardProps = {
  lessonId: number
  starGiven: string
  imageSrc: string
  reviewUrl: string
  chatUrl: string
}

export const ChallengesCompletedCard: React.FC<ChallengesCompletedCardProps> = ({
  imageSrc,
  chatUrl,
  reviewUrl,
  starGiven,
  lessonId
}) => {
  const [show, setShow] = useState<boolean>(false)
  const [star, setStar] = useState<string>(starGiven)
  return (
    <>
      <GiveStarCard
        close={() => setShow(false)}
        show={show}
        starGiven={star}
        setStarGiven={setStar}
        lessonId={lessonId}
      />
      <div className="card text-center shadow-sm">
        <div className="card-body">
          <img src={`/assets/curriculum/icons/${imageSrc}`} className="mt-3" />
          <h4 className="card-title mt-2">Congratulations!</h4>
          <p className="success-message">
            You have successfully completed all challenges
          </p>
          <p className="review-message">
            You can help your peers by
            <NavLink path={chatUrl} className="font-weight-bold mx-1" external>
              answering questions
            </NavLink>
            they have in the lesson and
            <NavLink
              path={reviewUrl}
              className="font-weight-bold mx-1"
              external
            >
              reviewing challenge submissions
            </NavLink>
          </p>
        </div>
        <div className="card-footer d-flex bg-primary">
          <p className="text-white mr-3 my-2">
            You can show your appreciation to the user that helped you the most
            by giving them a star
          </p>
          <button
            className="btn btn-light text-primary font-weight-bold ml-auto"
            onClick={() => setShow(true)}
          >
            Give Star
          </button>
        </div>
      </div>
    </>
  )
}

type ChallengeMaterialProps = {
  challenges: Challenge[]
  userSubmissions: UserSubmission[]
  lessonStatus: LessonStatus
  chatUrl: string
  lessonId: number
}

const ChallengeMaterial: React.FC<ChallengeMaterialProps> = ({
  challenges,
  userSubmissions,
  lessonStatus,
  chatUrl,
  lessonId
}) => {
  if (!challenges.length) {
    return <h1>No Challenges for this lesson</h1>
  }
  //create an object to evaluate the student's status with a challenge
  const userSubmissionsObject: UserSubmissionsObject = userSubmissions.reduce(
    (acc: UserSubmissionsObject, submission: UserSubmission) => {
      acc[submission.challengeId] = submission
      return acc
    },
    {}
  )
  //chain sort-map causes errors in tests without a copy
  const challengesCopy = [...challenges]
  //create a new Challenges array with user submission data integrated and sorted
  const challengesWithSubmissionData: ChallengeSubmissionData[] = challengesCopy
    .sort((a, b) => a.order - b.order)
    .map((challenge: Challenge) => {
      const submission = userSubmissionsObject[challenge.id] || {}
      return {
        ...challenge,
        status: submission.status || 'unsubmitted',
        submission
      }
    })
  const [
    currentChallengeID,
    setCurrentChallenge
  ] = useState<CurrentChallengeID>(null)

  const finalChallenge = {
    title: 'Challenges Completed!',
    id: 'finalChallenge',
    order: challenges.length + 1,
    description:
      'Congratulations, you have completed all Challenges for this Lesson',
    status: 'passed'
  }
  //find first challenge that is not passed on initial render after clicks will render clicked challenge
  const currentChallenge =
    challengesWithSubmissionData.find((challenge: ChallengeSubmissionData) => {
      if (currentChallengeID) return challenge.id === currentChallengeID
      return challenge.status !== 'passed'
    }) || finalChallenge
  const challengeTitleCards: React.ReactElement[] = challengesWithSubmissionData.map(
    challenge => {
      return (
        <ChallengeTitleCard
          key={challenge.id}
          id={challenge.id}
          challengeNum={challenge.order}
          title={challenge.title}
          setCurrentChallenge={setCurrentChallenge}
          active={challenge.id === currentChallenge.id}
          submissionStatus={challenge.status}
        />
      )
    }
  )
  return (
    <div className="row challenge-display mt-3">
      <div className="col-4">
        {challengeTitleCards}
        {lessonStatus.isPassed && (
          <ChallengeTitleCard
            key={finalChallenge.id}
            id={finalChallenge.id}
            challengeNum={finalChallenge.order}
            title={finalChallenge.title}
            setCurrentChallenge={setCurrentChallenge}
            active={finalChallenge.id === currentChallenge.id}
            submissionStatus={finalChallenge.status}
          />
        )}
      </div>
      <div className="col-8">
        {currentChallenge.id !== 'finalChallenge' && (
          <ChallengeQuestionCard currentChallenge={currentChallenge} />
        )}
        {lessonStatus.isPassed && currentChallenge.id === 'finalChallenge' && (
          <ChallengesCompletedCard
            lessonId={lessonId}
            starGiven={lessonStatus.starGiven || ''}
            imageSrc="icon-challenge-complete.jpg"
            chatUrl={chatUrl}
            reviewUrl={`https://www.c0d3.com/review/${lessonId}`}
          />
        )}
      </div>
    </div>
  )
}

export default ChallengeMaterial
