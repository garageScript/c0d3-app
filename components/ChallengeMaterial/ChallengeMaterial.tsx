import React, {
  useState,
  useContext,
  Dispatch,
  SetStateAction,
  useEffect
} from 'react'
import {
  Challenge,
  Submission,
  UserLesson,
  useGetPreviousSubmissionsQuery
} from '../../graphql/index'
import NavLink from '../NavLink'
import dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime'
import LocalizedFormat from 'dayjs/plugin/localizedFormat'
import GiveStarCard from '../GiveStarCard'
import _ from 'lodash'
import Modal from 'react-bootstrap/Modal'
import { SubmissionStatus, useAddCommentMutation } from '../../graphql'
import DiffView from '../DiffView'
import { Button } from '../theme/Button'
import MdInput from '../MdInput'
import { updateCache } from '../../helpers/updateCache'
import { GlobalContext } from '../../helpers/globalContext'
import SubmissionComments from '../SubmissionComments'
import SelectIteration from '../SelectIteration'
import Error, { StatusCode } from '../Error'
import ReviewStatus from '../ReviewStatus'
import useBreakpoint from '../../helpers/useBreakpoint'
import HighlightMarkdown from '../mdx/HighlightMarkdown/HighlightMarkdown'
import { useRouter } from 'next/router'
import { CURRICULUM_PATH } from '../../constants'

dayjs.extend(relativeTime)
dayjs.extend(LocalizedFormat)

type CurrentChallengeID = number | null

type ChallengeSubmissionData = Challenge & {
  status: string
  submission?: Submission
}

type UserSubmissionsObject = {
  [submissionId: number]: Submission
}

type StatusIconProps = {
  status: string
}
const StatusIcon: React.FC<StatusIconProps> = ({ status }) => {
  if (status === 'unsubmitted') {
    return null
  }
  let statusIconUrl
  switch (status) {
    case SubmissionStatus.Passed:
      statusIconUrl = '/assets/curriculum/icons/checkmark.svg'
      break
    case SubmissionStatus.NeedMoreWork:
      statusIconUrl = '/assets/curriculum/icons/rejected.svg'
      break
    case SubmissionStatus.Open:
      statusIconUrl = '/assets/curriculum/icons/pending.svg'
  }
  return <img width="25px" height="25px" src={statusIconUrl} />
}

type ChallengeTitleCardProps = {
  key: number
  id: number
  title: string
  challengeNum: number
  submissionStatus: string
  active?: boolean
  setCurrentChallenge: React.Dispatch<CurrentChallengeID>
}

export const ChallengeTitleCard: React.FC<ChallengeTitleCardProps> = ({
  submissionStatus,
  active,
  title,
  id,
  challengeNum,
  setCurrentChallenge
}) => {
  const router = useRouter()
  const { lessonSlug } = router.query

  const cardStyles: string[] = ['challenge-title-card']
  if (active) {
    cardStyles.push('challenge-title-card--active')
  }
  if (submissionStatus === SubmissionStatus.Passed) {
    cardStyles.push('challenge-title-card--done')
  } else {
    cardStyles.push('shadow-sm', 'border-0')
  }

  const handleChallengeClick = () => {
    router.push(`${CURRICULUM_PATH}/${lessonSlug}?challenge=${id}`, undefined, {
      shallow: true
    })

    setCurrentChallenge(id)
  }

  return (
    <div
      data-testid="challenge-title"
      className={`card mb-2 ${cardStyles.join(' ')}`}
      onClick={handleChallengeClick}
    >
      <div className="card-body d-flex justify-content-between">
        <div>{`${challengeNum}. ${title}`}</div>
        <StatusIcon status={submissionStatus} />
      </div>
    </div>
  )
}

const ChallengeQuestionCardDisplay: React.FC<{
  currentChallenge: ChallengeSubmissionData
}> = ({ currentChallenge }) => {
  const context = useContext(GlobalContext)
  const name = context.session?.user?.name
  const username = context.session?.user?.username
  const userId = context.session?.user?.id

  const [index, setIndex] = useState(-1)

  const [submission, setSubmission] = useState(currentChallenge.submission)
  const comments = submission?.comments?.filter(comment => !comment.line)

  const [commentValue, setCommentValue] = React.useState('')
  const [addComment] = useAddCommentMutation()
  const { data, loading, error } = useGetPreviousSubmissionsQuery({
    variables: {
      challengeId: submission?.challengeId!,
      userId: userId!
    },
    skip: userId === undefined || submission?.challengeId === undefined
  })
  useEffect(() => {
    if (data?.getPreviousSubmissions) {
      if (data.getPreviousSubmissions[index])
        setSubmission(data.getPreviousSubmissions[index] as Submission)
      else setSubmission(_.last(data.getPreviousSubmissions) as Submission)
    }
  }, [data, index])
  useEffect(() => {
    setIndex(-1)
    setSubmission(currentChallenge.submission)
  }, [currentChallenge.submission?.challengeId])
  if (submission && Object.keys(submission).length) {
    if (!name || !username || !userId)
      return (
        <Error
          code={StatusCode.INTERNAL_SERVER_ERROR}
          message="Error while retrieving userinfo from context"
        />
      )

    const update = updateCache({
      submissionId: submission.id,
      content: commentValue,
      name,
      username,
      lessonId: submission.lessonId,
      challengeId: submission.challengeId,
      userId: submission.user.id
    })

    return (
      <div className="card shadow-sm border-0 mt-3">
        {submission.createdAt && (
          <div
            className="card-header bg-white"
            title={dayjs(Number.parseInt(submission.createdAt)).format('LLLL')}
          >
            Submitted {dayjs(parseInt(submission.createdAt)).fromNow()}
          </div>
        )}
        <div className="text-start ms-2">
          <SelectIteration
            data={data}
            loading={loading}
            error={error}
            setIndex={setIndex}
            setSubmission={
              setSubmission as Dispatch<SetStateAction<Submission>>
            }
            currentIndex={index}
            currentId={submission.id}
          />
        </div>
        <div className="card-body">
          <div className="rounded-3 overflow-hidden">
            <DiffView
              submission={submission}
              generalStatus={currentChallenge.submission?.status}
            />
            {comments && (
              <SubmissionComments comments={comments} submission={submission} />
            )}
            <ReviewStatus
              name={submission.reviewer?.name}
              username={submission.reviewer?.username}
              date={submission.updatedAt}
              comment={submission.comment}
              status={submission.status}
              viewedByStudent
            />
            {currentChallenge?.submission?.status === SubmissionStatus.Open && (
              <>
                <MdInput
                  onChange={setCommentValue}
                  bgColor={'white'}
                  value={commentValue}
                />
                <Button
                  m="1"
                  btnType="success"
                  color="white"
                  onClick={() => {
                    if (!commentValue) return

                    addComment({
                      variables: {
                        submissionId: submission!.id,
                        content: commentValue
                      },
                      update
                    })
                    setCommentValue('')
                  }}
                >
                  Comment
                </Button>
              </>
            )}
          </div>
        </div>
      </div>
    )
  }

  return <></>
}

type ChallengeQuestionCardProps = {
  currentChallenge: ChallengeSubmissionData
}
export const ChallengeQuestionCard: React.FC<ChallengeQuestionCardProps> = ({
  currentChallenge
}) => {
  return (
    <>
      <div className="card shadow-sm border-0">
        <div className="card-body challenge-question">
          <h1 data-testid="challenge-question-title" className="card-title">
            {currentChallenge.title}
          </h1>
          <div
            data-testid="challenge-question-description"
            className="pt-3 pb-3 mt-3"
          >
            <HighlightMarkdown>
              {currentChallenge.description}
            </HighlightMarkdown>
          </div>
        </div>
      </div>

      <ChallengeQuestionCardDisplay currentChallenge={currentChallenge} />
    </>
  )
}

type ChallengesCompletedCardProps = {
  lessonId: number
  starGiven: string
  imageSrc: string
  reviewUrl: string
  chatUrl: string
}

export const ChallengesCompletedCard: React.FC<
  ChallengesCompletedCardProps
> = ({ imageSrc, chatUrl, reviewUrl, starGiven, lessonId }) => {
  const [show, setShow] = useState<boolean>(false)
  const [star, setStar] = useState<string>(starGiven)
  return (
    <>
      <GiveStarCard
        close={() => setShow(false)}
        show={show}
        starGiven={star}
        setStarGiven={setStar}
        lessonId={lessonId}
      />
      <div className="card text-center shadow-sm">
        <div className="card-body">
          <img src={`/assets/curriculum/icons/${imageSrc}`} className="mt-3" />
          <h4 className="card-title mt-2">Congratulations!</h4>
          <p className="success-message">
            You have successfully completed all challenges
          </p>
          <p className="review-message">
            You can help your peers by
            <NavLink path={chatUrl} className="fw-bold mx-1" external>
              answering questions
            </NavLink>
            they have in the lesson and
            <NavLink path={reviewUrl} className="fw-bold mx-1" external>
              reviewing challenge submissions
            </NavLink>
          </p>
        </div>
        <div className="card-footer d-flex bg-primary">
          <p className="text-white me-3 my-2">
            You can show your appreciation to the user that helped you the most
            by giving them a star
          </p>
          <button
            className="btn btn-light text-primary fw-bold ms-auto"
            onClick={() => setShow(true)}
          >
            Give Star
          </button>
        </div>
      </div>
    </>
  )
}

type ChallengeMaterialProps = {
  challenges: Challenge[]
  userSubmissions: Submission[]
  lessonStatus: UserLesson
  chatUrl: string
  lessonId: number
  lessonSlug: string
  show: boolean
  setShow: React.Dispatch<React.SetStateAction<boolean>>
}

const ChallengeMaterial: React.FC<ChallengeMaterialProps> = ({
  challenges,
  userSubmissions,
  lessonStatus,
  chatUrl,
  lessonId,
  lessonSlug,
  show,
  setShow
}) => {
  const isSmallDownBreakpoint = useBreakpoint('sm', 'down')
  if (!challenges.length) {
    return <h1>No Challenges for this lesson</h1>
  }

  const router = useRouter()
  const { challenge } = router.query

  //create an object to evaluate the student's status with a challenge
  const userSubmissionsObject: UserSubmissionsObject = userSubmissions.reduce(
    (acc: UserSubmissionsObject, submission: Submission) => {
      acc[submission.challengeId] = submission
      return acc
    },
    {}
  )
  //chain sort-map causes errors in tests without a copy
  const challengesCopy = [...challenges]
  //create a new Challenges array with user submission data integrated and sorted
  const challengesWithSubmissionData: ChallengeSubmissionData[] = challengesCopy
    .sort((a, b) => a.order - b.order)
    .map((challenge: Challenge) => {
      const submission = userSubmissionsObject[challenge.id] || {}
      return {
        ...challenge,
        status: submission.status || 'unsubmitted',
        submission
      }
    })
  const [currentChallengeID, setCurrentChallenge] =
    useState<CurrentChallengeID>(_.toNumber(challenge))

  const finalChallenge = {
    title: 'Challenges Completed!',
    id: 0,
    order: challenges.length + 1,
    lessonId: 0,
    description:
      'Congratulations, you have completed all Challenges for this Lesson',
    status: SubmissionStatus.Passed
  }
  //find first challenge that is not passed on initial render after clicks will render clicked challenge
  const currentChallenge =
    challengesWithSubmissionData.find((challenge: ChallengeSubmissionData) => {
      if (currentChallengeID) return challenge.id === currentChallengeID
      return challenge.status !== SubmissionStatus.Passed
    }) || finalChallenge
  const challengeTitleCards: React.ReactElement[] =
    challengesWithSubmissionData.map(challenge => {
      return (
        <ChallengeTitleCard
          key={challenge.id}
          id={challenge.id}
          challengeNum={challenge.order}
          title={challenge.title}
          setCurrentChallenge={setCurrentChallenge}
          active={challenge.id === currentChallenge.id}
          submissionStatus={challenge.status}
        />
      )
    })
  const challengeList = (
    <div
      className={`challenge-display_challenges col-md-4`}
      onClick={() => {
        setShow(!show)
      }}
    >
      {challengeTitleCards}
      {lessonStatus.passedAt && (
        <ChallengeTitleCard
          key={finalChallenge.id}
          id={finalChallenge.id}
          challengeNum={finalChallenge.order}
          title={finalChallenge.title}
          setCurrentChallenge={setCurrentChallenge}
          active={finalChallenge.id === currentChallenge.id}
          submissionStatus={finalChallenge.status}
        />
      )}
    </div>
  )
  return (
    <div className="row challenge-display mt-3">
      {isSmallDownBreakpoint ? (
        <Modal
          show={show}
          onHide={() => {
            setShow(!show)
          }}
          centered
          className="challenge-modal"
          data-testid="modal-challenges"
        >
          <Modal.Body>{challengeList}</Modal.Body>
        </Modal>
      ) : (
        challengeList
      )}

      <div className="col-md-8">
        {currentChallenge.id !== finalChallenge.id && (
          <ChallengeQuestionCard currentChallenge={currentChallenge} />
        )}
        {lessonStatus.passedAt && currentChallenge.id === 0 && (
          <ChallengesCompletedCard
            lessonId={lessonId}
            starGiven={lessonStatus.starGiven || ''}
            imageSrc="icon-challenge-complete.jpg"
            chatUrl={chatUrl}
            reviewUrl={`https://www.c0d3.com/review/${lessonSlug}`}
          />
        )}
      </div>
    </div>
  )
}

export default ChallengeMaterial
