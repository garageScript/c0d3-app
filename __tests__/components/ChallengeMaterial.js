import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import ChallengeMaterial from '../../components/ChallengeMaterial'

describe('Curriculum challenge page', () => {
  test('Should render appropriately when no challenges are passed to component', async () => {
    const { container } = render(<ChallengeMaterial />)
    await wait(() => {
      expect(container).toMatchSnapshot()
    })
  })

  const props = {
    challenges: [
      {
        id: '105',
        description:
          'Write a function that takes in a number and returns true if that number is greater than 5. Otherwise, return false.',
        title: 'Greater than 5',
        order: 0,
        __typename: 'Challenge'
      },
      {
        id: '107',
        description:
          'Write a function that takes in 2 numbers and returns their sum.',
        title: 'Sum of 2 Numbers',
        order: 1,
        __typename: 'Challenge'
      }
    ]
  }
  test('Should render first challenge by default', async () => {
    const { container } = render(<ChallengeMaterial {...props} />)
    expect(container).toMatchSnapshot()
  })

  test('Should render clicked challenge within challenge question', async () => {
    const { getAllByTestId, container } = render(
      <ChallengeMaterial {...props} />
    )
    const challengeTitleCard = getAllByTestId('challenge-title')[1]
    await wait(() => {
      fireEvent.click(challengeTitleCard), expect(container).toMatchSnapshot()
    })
  })
})
