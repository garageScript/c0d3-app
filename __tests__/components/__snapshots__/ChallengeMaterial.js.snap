// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Curriculum challenge page Should render appropriately when no challenges are passed to component 1`] = `
<div>
  <h1>
    No Challenges for this lesson
  </h1>
</div>
`;

exports[`Curriculum challenge page Should render clicked challenge within challenge question 1`] = `
<div>
  <div
    class="row challenge-display"
  >
    <div
      class="col-4"
    >
      <div
        class="card shadow-sm border-0 mb-2 challenge-title"
        data-testid="challenge-title"
      >
        <div
          class="card-body"
        >
          <div>
            0. Greater than 5
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm border-0 mb-2 challenge-title"
        data-testid="challenge-title"
      >
        <div
          class="card-body"
        >
          <div>
            1. Sum of 2 Numbers
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-8"
    >
      <div
        class="card shadow-sm border-0"
      >
        <div
          class="card-body challenge-question"
        >
          <h1
            class="card-title"
            data-testid="challenge-question-title"
          >
            Sum of 2 Numbers
          </h1>
          <div>
            <p
              class="card-question bg-light p-3 mt-3"
              data-testid="challenge-question-description"
            >
              Write a function that takes in 2 numbers and returns their sum.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Curriculum challenge page Should render first challenge by default 1`] = `
<div>
  <div
    class="row challenge-display"
  >
    <div
      class="col-4"
    >
      <div
        class="card shadow-sm border-0 mb-2 challenge-title"
        data-testid="challenge-title"
      >
        <div
          class="card-body"
        >
          <div>
            0. Greater than 5
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm border-0 mb-2 challenge-title"
        data-testid="challenge-title"
      >
        <div
          class="card-body"
        >
          <div>
            1. Sum of 2 Numbers
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-8"
    >
      <div
        class="card shadow-sm border-0"
      >
        <div
          class="card-body challenge-question"
        >
          <h1
            class="card-title"
            data-testid="challenge-question-title"
          >
            Greater than 5
          </h1>
          <div>
            <p
              class="card-question bg-light p-3 mt-3"
              data-testid="challenge-question-description"
            >
              Write a function that takes in a number and returns true if that number is greater than 5. Otherwise, return false.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
